packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js

```diff
@@ -170,24 +179,124 @@ export default class SidebarAdapter {
 			'',
 			settingArgs
 		);
-		setting.set( {} );
+		setting.set( widget.instance );
 
-		const widgetIds = [ ...this.setting.get() ];
 		const widgetId = settingIdToWidgetId( settingId );
-		widgetIds.splice( index, 0, widgetId );
-		this.setting.set( widgetIds );
 
 		return widgetId;
 	}
 
-	updateWidget( widget ) {
+	_updateWidget( widget ) {
+		const prevWidget = this.getWidget( widget.id );
+
+		// Bail out update.
+		if ( prevWidget === widget ) {
+			return widget.id;
+		}
+
 		const settingId = widgetIdToSettingId( widget.id );
 		this.allSettings( settingId ).set( widget.instance );
 		// TODO: what about the other stuff?
+
+		return widget.id;
 	}
 
-	removeWidget( widgetId ) {
-		const widgetIds = this.setting.get();
-		this.setting.set( without( widgetIds, widgetId ) );
+	getWidget( widgetId ) {
+		if ( ! widgetId ) {
+			return null;
+		}
+
+		const { idBase, number } = parseWidgetId( widgetId );
+		const settingId = widgetIdToSettingId( widgetId );
+		const setting = this.allSettings( settingId );
+
+		if ( ! setting ) {
+			return null;
+		}
+
+		const instance = setting.get();
+
+		if ( this.widgetsMap.has( instance ) ) {
+			return this.widgetsMap.get( instance );
+		}
+
+		const widget = {
+			id: widgetId,
+			idBase,
+			number,
+			instance,
+		};
+
+		this.widgetsMap.set( instance, widget );
+
+		return widget;
+	}
+
+	_internalUpdateWidgets( nextWidgets ) {
```

Comment:
I don't think we need to say `internal`. The `_` prefix makes that clear enough.